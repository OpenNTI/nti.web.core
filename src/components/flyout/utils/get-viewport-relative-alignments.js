import { getElementRect as getRectInViewport } from '@nti/lib-dom';

/**
 * @param  {HTMLElement} element   A dom node to measure
 * @param  {Object}  alignment The alignment description
 * @param  {DOMRect}    viewport  The viewport rect.
 * @returns {DOMRect} A partial rect.
 */
export function getViewportRelativeAlignments(element, alignment, viewport) {
	//the alignment is relative to the coordinateRoot. We need to constrain to the screen...
	//so we need to get the current screen coordinates.
	const rect = getRectInViewport(element);

	// We now have viewport relative rect, but the alignments omit keys that do not apply...
	// so we must also omit those keys.

	/* istanbul ignore else */
	if (alignment.top != null) {
		delete rect.bottom;
	}
	/* istanbul ignore else */
	if (alignment.left != null) {
		delete rect.right;
	}
	/* istanbul ignore else */
	if (alignment.bottom != null) {
		delete rect.top;
	}
	/* istanbul ignore else */
	if (alignment.right != null) {
		delete rect.left;
	}

	// ClientRects left & bottom's are distance from 0,0 (top, left), where
	// "css" bottom & left are the distance from the bottom & left sides..so we have to flip here.
	/* istanbul ignore else */
	if (rect.bottom != null) {
		rect.bottom = viewport.height - rect.bottom;
	}
	/* istanbul ignore else */
	if (rect.right != null) {
		rect.right = viewport.width - rect.right;
	}

	return rect;
}
