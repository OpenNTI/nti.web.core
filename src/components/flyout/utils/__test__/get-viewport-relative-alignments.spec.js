/* eslint-env jest */

import { getViewportRelativeAlignments } from '../get-viewport-relative-alignments';

describe('getViewportRelativeAlignments', () => {
	test('getViewportRelativeAlignments', () => {
		const viewport = { width: 1024, height: 768 };

		document.body.innerHTML = `
			<div>
				<div id=target></div>
			</div>
		`;
		const div = document.getElementById('target');
		Object.defineProperty(div, 'getBoundingClientRect', {
			value: () => ({
				top: 100,
				left: 50,
				bottom: 700,
				right: 250,
				width: 200,
				height: 600,
			}),
		});

		expect(
			getViewportRelativeAlignments(div, {}, viewport)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(
				div,
				{ top: 1, left: 1, right: 1, bottom: 1 },
				viewport
			)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(div, { top: 1, left: 1 }, viewport)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(
				div,
				{ bottom: 1, right: 1 },
				viewport
			)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(div, { top: 1 }, viewport)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(div, { bottom: 1 }, viewport)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(div, { left: 1 }, viewport)
		).toMatchSnapshot();
		expect(
			getViewportRelativeAlignments(div, { right: 1 }, viewport)
		).toMatchSnapshot();
	});
});
