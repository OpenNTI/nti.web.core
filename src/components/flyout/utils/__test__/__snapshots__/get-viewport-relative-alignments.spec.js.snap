// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 1`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 2`] = `
Object {
  "height": 600,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 3`] = `
Object {
  "height": 600,
  "left": 50,
  "top": 100,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 4`] = `
Object {
  "bottom": 68,
  "height": 600,
  "right": 774,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 5`] = `
Object {
  "height": 600,
  "left": 50,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 6`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "right": 774,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 7`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "top": 100,
  "width": 200,
}
`;

exports[`getViewportRelativeAlignments getViewportRelativeAlignments 8`] = `
Object {
  "bottom": 68,
  "height": 600,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;
